File: java/ToolGood/Words/internals/TrieNode2.java
package ToolGood.Words.internals;

import java.util.Hashtable;
import java.util.Map;

public class TrieNode2{
    public boolean End=false;
    public Map<String, Integer> Results;
    public Map<Character, TrieNode2> m_values;
    private int minflag = Integer.MAX_VALUE;
    private int maxflag = Integer.MIN_VALUE;


    public TrieNode2()
    {
        m_values = new Hashtable<Character, TrieNode2>();
        Results = new Hashtable<String, Integer>();
    }

    public boolean HasKey(Character c)
    {
        if (minflag <= c && maxflag >= c) {
            return m_values.containsKey(c);
        }
        return false;
    }
    public TrieNode2 GetValue(Character c ){
        return m_values.get(c);
    }

    public TrieNode2 Add(char c)
    {
        if( m_values.containsKey(c)){
            return m_values.get(c);
        }

        if (minflag > c) { minflag = c; }
        if (maxflag < c) { maxflag = c; }
        TrieNode2 node  = new TrieNode2();
        m_values.put(c, node);
        return node;
    }
 

    public void SetResults(String text, int index)
    {
        if (End == false) {
            End = true;
        }
        Results.put(text, index);
    }

    public void Merge(TrieNode2 node, Map<TrieNode2, TrieNode2> links)
    {
        if (node.End==true) {
            if (End == false) {
                End = true;
            }
            node.Results.forEach((key,Value)->{
                if (Results.containsKey(key) == false) {
                    Results.put(key,Value);
                }
            });
        }
        node.m_values.forEach((Key,Value)->{
            if ( m_values.containsKey(Key) == false) {
                if (minflag > Key) { minflag = Key; }
                if (maxflag < Key) { maxflag = Key; }
                m_values.put(Key, Value);
            }
        });
        if(links.containsKey(node)){
            TrieNode2 node2= links.get(node);
            Merge(node2, links);
        }
    }

 

}

File: golang/ToolGood/Words/internals/TrieNode2.go
package internals

type TrieNode2 struct {
	End      bool
	Results  []int
	M_values map[int32]*TrieNode2
	minflag  int32
	maxflag  int32
}

func NewTrieNode2() *TrieNode2 {
	return &TrieNode2{
		End:      false,
		M_values: make(map[int32]*TrieNode2),
		Results:  make([]int, 0),
		minflag:  0,
		maxflag:  0xffff,
	}
}

func (this *TrieNode2) Add(c int32, node *TrieNode2) {
	if this.minflag < c {
		this.minflag = c
	}
	if this.maxflag > c {
		this.maxflag = c
	}
	this.M_values[c] = node
}

func (this *TrieNode2) SetResults(text int) {
	if this.End == false {
		this.End = true
	}
	for i := 0; i < len(this.Results); i++ {
		if this.Results[i] == text {
			return
		}
	}
	this.Results = append(this.Results, text)
}

func (this *TrieNode2) TryGetValue(c int32) (bool, *TrieNode2) {
	if this.minflag <= c && this.maxflag >= c {
		if val, s := this.M_values[c]; s {
			return true, val
		}
	}
	return false, nil
}

func (this *TrieNode2) HasKey(c int32) bool {
	if this.minflag <= c && this.maxflag >= c {
		if _, s := this.M_values[c]; s {
			return true
		}
	}
	return false
}
